src/mscorlib/shared/System/IO/Path.Windows.cs

```diff
@@ -87,6 +87,54 @@ public static string GetFullPath(string path)
             return fullPath;
         }
 
+        public static string GetFullPath(string path, string basePath)
+        {
+            if (path == null)
+                throw new ArgumentException(nameof(path));
+
+            if (basePath == null)
+                throw new ArgumentNullException(nameof(basePath));
+
+            if (!IsPathFullyQualified(basePath))
+                throw new ArgumentException(SR.Arg_BasePathNotFullyQualified);
+
+            if (basePath.Contains('\0') || path.Contains('\0'))
+                throw new ArgumentException(SR.Argument_InvalidPathChars);
+
+            if (IsPathFullyQualified(path))
+                return GetFullPath(path);
+
+            int length = path.Length;
+
+            if (PathInternal.IsDevice(basePath))
+            {
+                // paths starting with \\?\ & \\.\
+                return RemoveRelativeSegments(CombineNoChecks(basePath, path), PathInternal.GetRootLength(basePath));
```

Comment:
This case is `Foo` and `\\?\C:\Bar`. You're returning `\\?\C:\Bar\Foo`. If someone passes in `C:Foo` or `\Foo` and `\\?\C:\Bar` we should probably handle that as well. The better way to handle this is probably to flip the logic around a bit.

``` C#
            string combinedPath = null;
            if ((length >= 1 && PathInternal.IsDirectorySeparator(path[0])))
            {
                // Path is current drive rooted i.e. starts with \:
                //   "\Foo" and "C:\Bar" => "C:\Foo"
                //   "\Foo" and "\\?\C:\Bar" => "\\?\C:\Foo"
                combinedPath = CombineNoChecks(GetPathRoot(basePath), path.AsReadOnlySpan().Slice(1));
            }
            else if (length >= 2 && PathInternal.IsValidDriveChar(path[0]) && path[1] == PathInternal.VolumeSeparatorChar)
            {
                // Drive relative paths
                Debug.Assert(length == 2 || !PathInternal.IsDirectorySeparator(path[2]));

                if (GetPathRoot(path.AsReadOnlySpan()) == GetPathRoot(basePath.AsReadOnlySpan()))
                {
                    // Matching root
                    //   "C:Foo" and "C:\Bar" => "C:\Bar\Foo"
                    //   "C:Foo" and "\\?\C:\Bar" => "\\?\C:\Bar\Foo"
                    combinedPath = CombineNoChecks(basePath, path.AsReadOnlySpan().Slice(2));
                }
                else
                {
                    // No matching root, root to specified drive
                    //   "D:Foo" and "C:\Bar" => "D:Foo"
                    //   "D:Foo" and "\\?\C:\Bar" => "\\?\C:\Bar"
                    combinedPath = path.Insert(2, "\\");
                }
            }
            else
            {
                // "Simple" relative path
                //   "Foo" and "C:\Bar" => "C:\Bar\Foo"
                //   "Foo" and "\\?\C:\Bar" => "\\?\C:\Bar\Foo"
                combinedPath = CombineNoChecks(basePath, path);
            }

            // Device paths are normalized by definition, so passing something of this format
            // to GetFullPath() won't do anything by design. Additionally, GetFullPathName() in
            // Windows doesn't root them properly. As such we need to manually remove segments.
            return PathInternal.IsDevice(combinedPath)
                ? RemoveRelativeSegments(combinedPath, PathInternal.GetRootLength(combinedPath))
                : GetFullPath(combinedPath);
```